<h1>Исследование надежности заемщиков<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Общая-информация" data-toc-modified-id="Шаг-1.-Общая-информация-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Общая информация</a></span></li><li><span><a href="#Шаг-2.-Предобработка-данных" data-toc-modified-id="Шаг-2.-Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2. Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Шаг-3.-Ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Ответьте-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 3. Ответьте на вопросы</a></span></li><li><span><a href="#Шаг-4.-Общий-вывод" data-toc-modified-id="Шаг-4.-Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Шаг 4. Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Шаг 1. Общая информация

Для начала посмотрим на общий вид таблицы - выведем первые 20 строк

In [1]:
import pandas as pd
df = pd.read_csv('/datasets/data.csv')
df.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Теперь посмотрим информацию о типах данных в таблице

Столбцы education и education_id, family_status и family_status_id должны быть связаны. Проверим - надо посчитать количество уникальных значений в этих парах и сравнить результаты - они должны совпадать. Так как в строке education разное форматирование (есть "среднее", "СРЕДНЕЕ" и "Среднее"), то ее надо проверять только после предобработки. Пока проверим семейный статус

In [2]:
#проверим семейный статус
print(df['family_status'].value_counts())
df['family_status_id'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64


0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Как и предполагалось - значения совпадают. Можно будет вынести family_status в отдельный словарь из основной таблицы, чтоб ее было легче читать (как и в случае education после проверки, что они реально совпадают) 

В столбцах "days_employed" и "total_income" одинаковое количество пропусков, они должны быть напрямую связаны и в одних и тех же строках. Убедимся в этом, выведем первые 10 пустых строк

In [3]:
df[df['days_employed'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Для ответа на поставленные вопросы нам нужны будут группировки по количеству детей ('children'), семейному положению ('family_status'), доходу ('total_income') и цели ('purpose'). Столбец семейное положение посмотрели выше, столбец целей нужно будет обрабатывать специальным образом.
    Посмотрим на значения столбцов 'children' и 'total_income' 

In [4]:
#посмотрим на крайние значения столбцов
print('Минимальное значение children:', df['children'].min())
print('Максимальное значение children:', df['children'].max())
print('Минимальное значение total_income:', df['total_income'].min())
print('Максимальное значение total_income:', df['total_income'].max())

Минимальное значение children: -1
Максимальное значение children: 20
Минимальное значение total_income: 20667.26379327158
Максимальное значение total_income: 2265604.028722744


Значения дохода в общем вопросов не вызывают, а в значении столбца 'children' явно есть ошибки - не может быть отрицательного количества детей. 20 детей - тоже довольно нетривиальное значение. В теории такое может быть, к примеру, приемные дети, но очень маловероятно. Проверим количество отрицательных значений, максимальных и вообще больших для данного столбца. 

In [5]:
print('Отрицательных значений:', df[df['children'] < 0]['children'].count())
print('20 детей', df[df['children'] == 20]['children'].count())
print('Больше 5 детей:', df[df['children'] > 5]['children'].count())

Отрицательных значений: 47
20 детей 76
Больше 5 детей: 76


- 47 строк имеют значение -1 в столбце "children". Это ничтожно мало от общего количество, но может помешать нам при группировке и ответах на вопросы. Самое очевидное объяснение - это просто неправильно внесенные значения "1". Возьмем эти значения по модулю.
- Поскольку 76 человек с 20 детьми - явно аномальное значение, особенно учитывая, что в остальных значениях - цифры не больше 5. Вероятнее всего это просто неправильно записанное значение 2. Надо вернуть его

**Вывод**

Из выведенных 20 строк видно:
- очень странные цифры трудового стажа в днях - есть отрицательные значения, сами значения - не целые, хотя "дни" вроде бы это подразумевают. На первый взгляд отрицательные значения по модулю на порядок меньше положительных;
- в столбце education - разное форматирование данных: как строчные, так и заглавные группы, нужно будет привести к одному виду.
    
Из информации о таблице:
- есть пропуски в столбцах "days_employed" и "total_income". Количество значений одинаковое, категории напрямую связаны - это связанные пропуски в одних и тех же строках. Очевидно, что для людей, у которых не было данных о стаже - не было данных и о среднем размере дохода;
- странный для значения в днях формат float столбца days_employed. Но и сами значения странные - надо будет разбираться.



Есть ошибки в столбце 'children'.

    

In [6]:
#посмотрим какие значения есть у столбца gender
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Одно значение явно аномальное (XNA). Посмотрим на всю строку

In [7]:
df[df['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0   -2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Вероятно, это какая-то ошибка при написании NaN. Можно было бы заменить на реальный NaN, но зачем плодить лишний пропуск. Можно заменить на любое из двух значений - вы выборке из 14 и 7 тыс.значений еще одно роли не сыграет `df.loc[10701, 'gender'] = 'M'`. Также можно было бы просто удалить строку, но для наших целей группировка по этому столбцу не нужна, а данные этой строки пригодятся. Поэтому я бы просто оставил его как есть.

Переходим к income_type

In [8]:
print(df['income_type'].describe())
df['income_type'].value_counts()

count         21525
unique            8
top       сотрудник
freq          11119
Name: income_type, dtype: object


сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

К income_type вопросов, кажется, нет, посмотрим dob_years

In [9]:
print(df['dob_years'].describe())
df['dob_years'].value_counts()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64


35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Есть 101 аномальное значение в 0 лет. Скорее это такой вид 'нет данных'

## Шаг 2. Предобработка данных

### Обработка пропусков

Прежде чем приступить к пропускам, разберемся с ошибками в столбце с количестовм детей. Возьмем значения - 1 по модулю и заменим 20 на 2

In [10]:
#убираем минус из столбца
df.loc[df['children'] == -1, 'children'] = df.loc[df['children'] == -1, 'children'].abs()
#превращаем 20 в 2
df.loc[df['children'] == 20, 'children'] = 2
#проверяем
print('Отрицательных значений:', df[df['children'] < 0]['children'].count())
print('20 детей', df[df['children'] == 20]['children'].count())
print('Больше 5 детей:', df[df['children'] > 5]['children'].count())

Отрицательных значений: 0
20 детей 0
Больше 5 детей: 0


Пропуски присутствуют в столбцах days_employed и total_income. Но сами данные в days_employed странные. Прежде чем удалять пропуски, нужно разобраться с этим столбцом

In [11]:
#Явно в days_employed есть два типа значений - положительные и отрицательные. Посчитаем сколько их:
print('Отрицательных значений:', df[df['days_employed'] < 0]['days_employed'].count())
print('Положительных значений:', df[df['days_employed'] > 0]['days_employed'].count())

Отрицательных значений: 15906
Положительных значений: 3445


Число значений значительное - посмотрим в каких они диапазонах - найдем максимальное и минимальное значение для каждой из групп, проверим, есть ли нулевой стаж

In [12]:

print('Минимум < 0:', df[df['days_employed'] < 0]['days_employed'].min())
print('Максимум < 0:', df[df['days_employed'] < 0]['days_employed'].max())
print('Минимум > 0:', df[df['days_employed'] > 0]['days_employed'].min())
print('Максимум > 0:', df[df['days_employed'] > 0]['days_employed'].max())
print('Стаж равен 0:', df[df['days_employed'] == 0]['days_employed'].max())

Минимум < 0: -18388.949900568383
Максимум < 0: -24.14163324048118
Минимум > 0: 328728.72060451825
Максимум > 0: 401755.40047533
Стаж равен 0: nan


Нулевых значений нет. Значения больше нуля на порядок по модулю отличаются от значений меньше нуля. При этом значения больше нуля слишком большие.  Если это дни, то разделим на количество дней в году даже минимальное значение:

In [13]:
df[df['days_employed'] > 0]['days_employed'].min() / 365

900.6266317932007

900 лет работы без выходных. Значение слишком большое. Допустим - это часы. Разделим на средний 8-часовой рабочий день, а затем снова на количество дней в году

In [14]:
df[df['days_employed'] > 0]['days_employed'].min() / 8 /365

112.57832897415008

112 лет - тоже слишком много. Тогда допустим, что часы посчитаны по астрономическим часам в сутках. Найдем минимальные и максимальные значения

In [15]:
print('Минимум > 0:', df[df['days_employed'] > 0]['days_employed'].min()/24/365)
print('Максимум > 0:', df[df['days_employed'] > 0]['days_employed'].max()/24/365)

Минимум > 0: 37.526109658050025
Максимум > 0: 45.8624886387363


Вот эти значения стажа подходят - это самое вероятное решение. Осталось привести данные к одному виду - взять по модулю отрицательные значения и перевести часы в дни

In [16]:
#сначала переводим часы (положительные значения) в дни
df.loc[df['days_employed'] > 0, 'days_employed'] = df['days_employed'] / 24
# теперь убираем минус из значений
df['days_employed'] = df['days_employed'].abs()
# проверяем
df.head(20)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0   14177.753002         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0   16678.380705         53              среднее             1   
19         0   10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Чтобы заполнить пропуски в 'days_employed', сгруппируем данные по 'income_type' и 'children'. Логика в следующем. Тип занятости может меняться с возрастом: пенсионер явно соответствует старшему возрасту, чтобы из сотрудника перерасти в компаньона (если я правильно понимаю что это) - тоже требуется время. Также и с рождением детей - наличие 3х детей указывает косвенно на больший возраст. Связи, конечно, опосредованные - предпринимателю может быть 18, а без детей можно прожить всю жизнь, но все равно это кажется несколько точнее, чем медиана по столбцу

In [17]:
#создадим фрейм с группированными значениями
group_age = df.groupby(['income_type', 'children']).agg({'dob_years': 'median'}).rename(columns = {'dob_years': 'med_dob_years'})
#дополняем исходный фрейм этими данными
df = df.merge(group_age, on = ['income_type', 'children'])
#проверяем, что получилась дополнительная колонка 'median_dob_years'
df[df['dob_years'] == 0].head(20)

children  days_employed  dob_years            education  education_id  \
853          1     162.936952          0              среднее             1   
963          1     731.396794          0               высшее             0   
966          1     260.992722          0  неоконченное высшее             2   
1263         1    1720.866870          0              среднее             1   
1441         1    1605.506161          0               высшее             0   
1464         1     908.172637          0              среднее             1   
1552         1     223.709545          0  неоконченное высшее             2   
1678         1    1015.395451          0              среднее             1   
1841         1    3208.732719          0               высшее             0   
1902         1    1018.525283          0              среднее             1   
2337         1    1524.869300          0              среднее             1   
2961         0    2664.273168          0              среднее             1   
3474         0            NaN          0               высшее             0   
3539         0   10689.250498          0              среднее             1   
3644         0     794.876230          0              среднее             1   
3648         0    2497.754414          0              среднее             1   
3989         0    4288.714410          0              среднее             1   
4107         0    3549.117333          0              среднее             1   
4939         0    1478.092467          0               высшее             0   
5046         0    1570.420460          0              среднее             1   

              family_status  family_status_id gender income_type  debt  \
853        гражданский брак                 1      F   сотрудник     0   
963               в разводе                 3      F   сотрудник     0   
966        гражданский брак                 1      F   сотрудник     0   
1263        женат / замужем                 0      F   сотрудник     0   
1441        женат / замужем                 0      F   сотрудник     0   
1464        женат / замужем                 0      M   сотрудник     0   
1552       гражданский брак                 1      M   сотрудник     0   
1678  Не женат / не замужем                 4      M   сотрудник     1   
1841       гражданский брак                 1      F   сотрудник     0   
1902        женат / замужем                 0      F   сотрудник     1   
2337        женат / замужем                 0      F   сотрудник     0   
2961              в разводе                 3      F   сотрудник     0   
3474  Не женат / не замужем                 4      F   сотрудник     0   
3539        женат / замужем                 0      F   сотрудник     0   
3644       гражданский брак                 1      F   сотрудник     0   
3648        женат / замужем                 0      F   сотрудник     1   
3989              в разводе                 3      F   сотрудник     0   
4107              в разводе                 3      M   сотрудник     0   
4939  Не женат / не замужем                 4      F   сотрудник     0   
5046       гражданский брак                 1      F   сотрудник     0   

       total_income                           purpose  med_dob_years  
853   106956.800302             на покупку автомобиля           37.0  
963   247540.528218                     покупка жилья           37.0  
966   128246.646591              покупка недвижимости           37.0  
1263  220298.348654            профильное образование           37.0  
1441  153478.301459  сделка с подержанным автомобилем           37.0  
1464  158458.734055                             жилье           37.0  
1552  256070.541695                   сыграть свадьбу           37.0  
1678  107342.751020                     покупка жилья           37.0  
1841   91924.300059             на проведение свадьбы           37.0  
1902  155341.706429                   свой автомобиль           37.0  
2337  1099

In [18]:
#переносим в нулевые ячейки dob_years медианные значения
df.loc[df['dob_years'] == 0, 'dob_years'] = df.loc[df['dob_years'] == 0, 'med_dob_years']
#проверяем
df.loc[df['dob_years'] == 0]['dob_years'].count()

0

In [19]:
#теперь собственно заполним аналогично стаж по возрасту и типу занятости
group_med_emp = (df.groupby(['income_type', 'dob_years']).agg({'days_employed': 'median'}).
                  rename(columns = {'days_employed': 'median_days_employed'}))
#дополняем исходный фрейм этими данными
df = df.merge(group_med_emp, on = ['dob_years', 'income_type'])
#проверяем таблицу
df.loc[df['days_employed'].isna()].head()

children  days_employed  dob_years education  education_id  \
18         1            NaN       42.0   среднее             1   
21         1            NaN       42.0   среднее             1   
32         1            NaN       42.0   среднее             1   
50         1            NaN       42.0   среднее             1   
51         1            NaN       42.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
18        в разводе                 3      F   сотрудник     0           NaN   
21  женат / замужем                 0      F   сотрудник     0           NaN   
32  женат / замужем                 0      F   сотрудник     1           NaN   
50  женат / замужем                 0      M   сотрудник     0           NaN   
51  женат / замужем                 0      M   сотрудник     0           NaN   

                                  purpose  med_dob_years  median_days_employed  
18                           недвижимость           37.0           1957.465867  
21               операции с недвижимостью           37.0           1957.465867  
32  получение дополнительного образования           37.0           1957.465867  
50                           недвижимость           37.0           1957.465867  
51                          покупка жилья           37.0           1957.465867

In [20]:
#переносим в пустые ячейки 'days_employed' значения 'median_days_employed'
df.loc[df['days_employed'].isna(), 'days_employed'] = df.loc[df['days_employed'].isna(), 'median_days_employed']
#проверяем пропуски
df.loc[df['days_employed'].isna()]

children  days_employed  dob_years education  education_id  \
14953         0            NaN       24.0   среднее             1   
20056         0            NaN       69.0    высшее             0   
21518         2            NaN       31.0   среднее             1   
21520         0            NaN       58.0    высшее             0   

               family_status  family_status_id gender      income_type  debt  \
14953        женат / замужем                 0      F        пенсионер     0   
20056  Не женат / не замужем                 4      F        компаньон     0   
21518       гражданский брак                 1      F        пенсионер     0   
21520        женат / замужем                 0      M  предприниматель     0   

       total_income                     purpose  med_dob_years  \
14953           NaN        покупка своего жилья           60.0   
20056           NaN                  автомобили           42.0   
21518           NaN                     свадьба           55.0   
21520           NaN  покупка жилой недвижимости           42.5   

       median_days_employed  
14953                   NaN  
20056                   NaN  
21518                   NaN  
21520                   NaN

Остались 4 пустых значения, которые, видимо, не были отнесены при группировке к одной из групп. Заполним медианой

In [22]:
df.loc[df['days_employed'].isna(), 'days_employed'] = df['days_employed'].median()
df['days_employed'].isna().sum()

0

Проводим аналогичные операции для стобца 'total_income'. На этот раз группируем по образованию и типу занятости

In [23]:
#создадим фрейм с группированными значениями
group_med_inc = (df.groupby(['income_type', 'education_id']).agg({'total_income': 'median'}).
                  rename(columns = {'total_income': 'median_total_income'}))
#дополняем исходный фрейм этими данными
df = df.merge(group_med_inc, on = ['income_type', 'education_id'])
#переносим в пустые ячейки 'total_income' значения 'median_total_income'
df.loc[df['total_income'].isna(), 'total_income'] = df.loc[df['total_income'].isna(), 'median_total_income']
#проверяем пропуски
df['total_income'].isna().sum()

0

**Вывод**

Данные в days_empolyed приведены к одному виду значений, все пропуски в таблице заполнены медианными значениями для тех групп, которые нам нужны для анализа.

### Замена типа данных

'days_employed' должен иметь целочисленные значения, раз измеряется в целых днях. Надо перевести значения столбца из float в int. Метод to_numeric() здесь не подходит, но подойдет astype()

In [24]:
#переводим в int
df['days_employed'] = df['days_employed'].astype(int)
df['total_income'] = df['total_income'].astype(int)
#проверяем
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21525 entries, 0 to 21524
Data columns (total 15 columns):
children                21525 non-null int64
days_employed           21525 non-null int64
dob_years               21525 non-null float64
education               21525 non-null object
education_id            21525 non-null int64
family_status           21525 non-null object
family_status_id        21525 non-null int64
gender                  21525 non-null object
income_type             21525 non-null object
debt                    21525 non-null int64
total_income            21525 non-null int64
purpose                 21525 non-null object
med_dob_years           21525 non-null float64
median_days_employed    21521 non-null float64
median_total_income     21525 non-null float64
dtypes: float64(4), int64(6), object(5)
memory usage: 2.6+ MB


**Вывод**

Все данные приведены к нужным типам

### Обработка дубликатов

Как было замечено еще на этапе изучения данных, в столбце 'education' есть разное написание значений. Посмотрим, какие тут варианты:

In [25]:
df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Очевидно все дело в регистре. Переведем все в нижний регистр.

In [26]:
df['education'] = df['education'].str.lower()
#Проверим результат и заодно проверим, что значения education и education_id совпадают. Посчитаем количество возможных значений,
#если они для двух столбцов совпадают, значит значения education в правильном регистре и education = education_id
print(df['education'].value_counts())
df['education_id'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64


1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

Теперь 'education' точно можно будет вынести в отдельный словарь

Теперь проверим, нет ли в таблице явных дубликатов. C одной стороны в таблице нет какого то уникального идентификатора, который бы позволил со 100% вероятностью утверждать, что перед нами дубликат, значения всех столбцов в приницпе могут дублироваться для разных строк. Но шанс того, что значения стажа в днях и среднего дохода до копеек, а также все остальные столбцы, совпадут - ничтожно мал. Вероятнее всего, это одни и те же люди, которые были почему-то записаны несколько раз. К примеру, подали заявку на кредит несколько раз, или просто произошел технический сбой. Поэтому дубликаты надо все же удалить.

In [27]:
df.duplicated().sum()

71

In [28]:
#удалим дубликаты
df = df.drop_duplicates().reset_index(drop=True)
#проверим результат
df.duplicated().sum()

0

**Вывод**

Данные в строке 'education' приведены к одному виду, найден и удален 71 явный дубликат.

### Лемматизация

Чтобы категоризировать заемщиков по целям кредита, надо эти цели как-то формализовать, разделить по категориям. Для начала посмотрим, какие они бывают в исходных данных:


In [29]:
df['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

Для выделения категорий лемматизируем значения

In [30]:
#выделим значения, склеим в одну строку и лемматизируем, подсчитаем количество упоминаний
from pymystem3 import Mystem
m = Mystem()
from collections import Counter


purp = df['purpose'].unique()
purp = ' '.join(purp)
purp = m.lemmatize(purp)
Counter(purp)

Counter({'покупка': 10,
         ' ': 96,
         'жилье': 7,
         'получение': 3,
         'образование': 9,
         'на': 4,
         'проведение': 1,
         'свадьба': 3,
         'сыграть': 1,
         'высокий': 3,
         'заниматься': 2,
         'недвижимость': 10,
         'строительство': 3,
         'собственный': 1,
         'дополнительный': 2,
         'операция': 4,
         'с': 5,
         'коммерческий': 2,
         'жилой': 2,
         'автомобиль': 9,
         'приобретение': 1,
         'для': 2,
         'семья': 1,
         'свой': 4,
         'сделка': 2,
         'подержать': 2,
         'профильный': 1,
         'ремонт': 1,
         'со': 1,
         'сдача': 1,
         '\n': 1})

**Вывод**

Очевидно, что есть некоторые категории: "свадьба", "автомобиль", "образование", не совсем ясно с категорией "недвижимость" - явно есть жилая, и коммерческая, и это, вероятно, очень разные категории по портретам заемщиков. Так же и с образованием - есть дополнительное и есть высшее - основное. Думаю, правильно будет выделять категории по следующим словам:
- "свадьба"
- "автомобиль"
- "коммерческий" и "недвижимость" (эта категория оченб мильно отличается от "жилье")
- "жилье" или "недвижимость" (если недвижимость коммерческая - об этом обычно указывают, поэтому просто "недвижимость" по умолчанию - к этой группе. Есть "жилье под сдачу" - но его трудно вычленить, подобные цели могут быть и просто в "жилье" и в "недвижимости". Учитывая, что это, зачастую, просто консервативный способ вложения денег для обычных людей, я бы не стал выделять в отдельную категорию)
- "дополнительный" и "образование"
- "образованиe" (помимо дополнительного - все высшее, профильное и т.д.)

### Категоризация данных

Во-первых, надо избавить данные от "дублирующих" столбцов 'family_status' и 'education' (вынести их в отдельные словари), а также убрать ненужные теперь 'median_days_employed' и 'median_total_income'.

In [31]:
family_dict=df[['family_status','family_status_id']]
family_dict = family_dict.drop_duplicates().reset_index(drop=True)
education_dict = df[['education','education_id']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True)
df = df.drop(['family_status', 'education', 'median_days_employed', 'median_total_income', 'med_dob_years'], axis = 1)
#проверим
print(family_dict)
print(education_dict)
df.head()

           family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2  Не женат / не замужем                 4
3              в разводе                 3
4         вдовец / вдова                 2
             education  education_id
0               высшее             0
1  неоконченное высшее             2
2              среднее             1
3            начальное             3
4       ученая степень             4


children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437       42.0             0                 0      F   
1         1           7215       42.0             0                 0      F   
2         1           4614       42.0             0                 0      M   
3         1            733       42.0             0                 1      M   
4         1           8899       42.0             0                 1      F   

  income_type  debt  total_income                purpose  
0   сотрудник     0        253875          покупка жилья  
1   сотрудник     0        143875  получение образования  
2   сотрудник     0       1278622          покупка жилья  
3   сотрудник     0        433203  на проведение свадьбы  
4   сотрудник     0        360887        сыграть свадьбу

Теперь создадим функцию, которая бы лемматизировала значения столбца 'purpose' и в соответствии с ними возвращала бы категорию займа и категоризируем с ее помощью заемщиков, добавив столбец 'category' в датасет

In [32]:
def categorize_purpose(row):
    row = m.lemmatize(row)
    if 'свадьба' in row:
        return 'wedding'
    elif 'автомобиль' in row:
        return 'car'
    elif 'коммерческий' in row and 'недвижимость' in row:
        return 'commercial'
    elif 'жилье' in row or 'недвижимость' in row:
        return 'real_estate'
    elif 'дополнительный' in row and 'образование' in row:
        return 'add_education'
    elif 'образование' in row:
        return 'gen_education'
    else:
        return 'unknown'
    
df['category'] = df['purpose'].apply(categorize_purpose)
#проверим
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437       42.0             0                 0      F   
1         1           7215       42.0             0                 0      F   
2         1           4614       42.0             0                 0      M   
3         1            733       42.0             0                 1      M   
4         1           8899       42.0             0                 1      F   

  income_type  debt  total_income                purpose       category  
0   сотрудник     0        253875          покупка жилья    real_estate  
1   сотрудник     0        143875  получение образования  gen_education  
2   сотрудник     0       1278622          покупка жилья    real_estate  
3   сотрудник     0        433203  на проведение свадьбы        wedding  
4   сотрудник     0        360887        сыграть свадьбу        wedding

Помимо целей кредита, нужно будет еще категоризировать заемщиков по урвоню дохода. Минимальное и максимальное значение нам известны:
- Минимальное значение total_income: 20667.26379327158
- Максимальное значение total_income: 2265604.028722744

Для этого надо написать функцию, аналогичную функции для целей, и применить ее к данным

In [33]:
def categorize_income(row, start, i, finish):
    if row >= finish:
        return (f'{finish} and more')
    else:
        while row >= start:
            start = start+i
        else:
            return (f'{start-i}-{start}')
        
#проверим
categorize_income(300000, 0, 30000, 300000)

'300000 and more'

In [35]:
#проверим сколько значений получилось
df['category_income'].value_counts()

120000-150000      4264
90000-120000       3822
150000-180000      3411
60000-90000        2542
180000-210000      2106
300000 and more    1483
210000-240000      1372
240000-270000       989
30000-60000         784
270000-300000       659
0-30000              22
Name: category_income, dtype: int64

**Вывод**

Данные очищены и категоризированы по целям кредита, можно приступать к ответу на вопросы.

## Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [36]:
#Упс! Сделал лишнее, но оставлю так
#сделаем группировку по количеству детей (столбцу 'children') и посчитаем 
#общее количество людей с таким количеством детей, людей, которые задолжали, 
#и cреднее значение - чем оно ближе к единице, тем больше процент должников
#children_debt = df.groupby(['children']).agg({'debt':['count','sum','mean']})
#children_debt = children_debt.rename(columns = {'children': 'количество детей', 'count': 'всего случаев',
                                              #'sum': 'должников','mean': 'доля должников'})
#children_debt

In [37]:
#посчитаем процент должников среди людей без детей и с детьми:
print('Доля должников среди людей без детей:', df[df['children'] == 0]['debt'].mean())
print('Доля должников среди людей с детьми:', df[df['children'] != 0]['debt'].mean())

Доля должников среди людей без детей: 0.07543822297920659
Доля должников среди людей с детьми: 0.09208203178052424


**Вывод**

Люди, у которых нет детей, возвращают кредиты лучше, чем люди с детьми. Должников среди них на почти на 2% меньше - 7,54% против 9,2% у людей с детьми

Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [38]:
#сгруппируем по family_status_id
family_debt = df.pivot_table(index=['family_status_id'], values = 'debt', aggfunc = ['count','sum','mean'])
#переименуем столбцы
family_debt.columns = ['всего случаев', 'должников', 'доля должников']
#"расшифруем" данные о семейном положении через присоеднинение family_dict
family_debt = family_debt.merge(family_dict, on = "family_status_id", how = 'left')
#уберем ненужный "family_status_id"
family_debt = family_debt.drop(['family_status_id'], axis = 1)
family_debt

всего случаев  должников  доля должников          family_status
0          12339        931        0.075452        женат / замужем
1           4151        388        0.093471       гражданский брак
2            959         63        0.065693         вдовец / вдова
3           1195         85        0.071130              в разводе
4           2810        274        0.097509  Не женат / не замужем

**Вывод**

- Семейное положение оказывает влияние на возврат кредита (разница процента должников в разных группах - до 3%).
- Меньше всего должников среди категории "вдовец / вдова" - 6,56%.
- Хуже всего возвращают кредиты неженатые люди (9,75% должников) и люди, живущие в гражданском браке (9,34% должников).
- Для самой многочисленной категории "женат / замужем" значения средние - 7,54% должников, чуть лучше показатель для людей в разводе - 7,11% должников.

Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [39]:
#сгруппируем по уровню дохода
#income_debt = df.groupby(['category_income']).agg({'debt':['count','sum','mean']})
income_debt = df.pivot_table(index=['category_income'], values = 'debt', aggfunc = ['count','sum','mean'])
#переименуем столбцы
income_debt.columns = ['всего случаев', 'должников', 'доля должников']
income_debt.sort_values(by = 'доля должников')

всего случаев  должников  доля должников
category_income                                          
30000-60000                784         47        0.059949
240000-270000              989         66        0.066734
270000-300000              659         45        0.068285
210000-240000             1372         98        0.071429
300000 and more           1483        106        0.071477
180000-210000             2106        171        0.081197
90000-120000              3822        319        0.083464
150000-180000             3411        285        0.083553
60000-90000               2542        213        0.083792
0-30000                     22          2        0.090909
120000-150000             4264        389        0.091229

**Вывод**

- Некоторая зависимость между урвонем доходи и возвратом кредита есть.
- Хуже всего возвращают кредиты люди с доходами в диапазоне 120 - 150 тыс. (9,12%) и с доходами до 30 тыс. (9,09% должников, впрочем, тут очень мало данных - 22 случая).
- Выделяется категория с доходом 30 - 60 тыс. - в ней рекордно низкий процент должников 5,99%. Но, несмотря на это в целом видна тенденция, что с ростом уровня дохода количество должников падает. Для категорий с доходом от 200 тыс. значения колеблятся около 7%, в то время как для категорий до 200 тыс. - это 8-9%.
- Разница не очень существенна, и, как мне кажется, здесь нужно проводить отдельное исследование - смотреть скорее на цели кредита в каждой из групп. Они должны быть разные в разных категориях и именно этим, вероятно, можно объяснить аномально низкий процент должников в группе 30-60 тыс. Скорее всего цель кредита и его объем, соотвественно, там гораздо меньше, чем в кредитах группы 150 - 180 тыс., например.  

Как разные цели кредита влияют на его возврат в срок?

In [40]:
#сгруппируем по уровню дохода
purpose_debt = df.groupby(['category']).agg({'debt':['count','sum','mean']})
#переименуем столбцы
purpose_debt = purpose_debt.rename(columns = {'count': 'всего случаев', 'sum': 'должников','mean': 'доля должников'})
purpose_debt

debt                         
              всего случаев должников доля должников
category                                            
add_education           906        89       0.098234
car                    4306       403       0.093590
commercial             1311        99       0.075515
gen_education          3107       281       0.090441
real_estate            9500       683       0.071895
wedding                2324       186       0.080034

**Вывод**

Есть прямая связь между целями кредитования и возвратом кредита в срок:
- Лучше всего возвращаются кредиты на недвижимость, причем на жилую чть лучше (7,18% должников), чем на коммерческую (7,55% должников)
- Хуже всего возвращаются кредиты на доп.образование (9,82% должников) и автомобили (9,35% должников)

## Шаг 4. Общий вывод

Каждый из показателей (наличие детей, семейное положение, цель кредита, уровень дохода) может указывать на вероятность его возврата в срок, увеличивая или уменьшая ее на 2-3%. Идеальный заемщик по этим показателям - вдова или вдовец без детей, взявший(ая) кредит на покупку жилья при высоком уровне дохода (от 210 тыс.). Худший из вариантов - одинокая мать с несколькими детьми, взявшая кредит на дополнительное образования. В данных об уровне дохода есть неравномерные показатели, которые, как мне кажется, говорят о необходимости изучения каждой из выделенных по уровню доход групп в разрезе целей кредитования.